packages/std/src/coin.rs

```diff
@@ -166,4 +182,41 @@ mod tests {
         // less than same type
         assert!(has_coins(&wallet, &coin(777, "ETH")));
     }
+
+    #[test]
+    fn parse_coin() {
+        let expected = Coin::new(123, "ucosm");
+        assert_eq!("123ucosm".parse::<Coin>().unwrap(), expected);
+        // leading zeroes should be ignored
+        assert_eq!("00123ucosm".parse::<Coin>().unwrap(), expected);
+        // 0 amount parses correctly
+        assert_eq!("0ucosm".parse::<Coin>().unwrap(), Coin::new(0, "ucosm"));
+        // ibc denom should work
+        let ibc_str = "11111ibc/27394FB092D2ECCD56123C74F36E4C1F926001CEADA9CA97EA622B25F41E5EB2";
+        let ibc_coin = Coin::new(
+            11111,
+            "ibc/27394FB092D2ECCD56123C74F36E4C1F926001CEADA9CA97EA622B25F41E5EB2",
+        );
+        assert_eq!(ibc_str.parse::<Coin>().unwrap(), ibc_coin);
+
+        // error cases
+        assert_eq!(
+            Coin::from_str("123").unwrap_err(),
+            CoinFromStrError::MissingDenom
+        );
+        assert_eq!(
+            Coin::from_str("ucosm").unwrap_err().to_string(),
+            "Invalid amount: cannot parse integer from empty string"
+        );
+        assert_eq!(
+            Coin::from_str("-123ucosm").unwrap_err().to_string(),
+            "Invalid amount: cannot parse integer from empty string"
+        );
```

Comment:
Interesting test cases here.

Can we have a dedicated check for cases where the number does not start with a digit? Otherwise the error message is more misleasing than helping. The basic idea would be
- Ensure we start reading digits (not yet implemented)
- Once we get a non-digit, mark as separation point to denom (already implemented)

This would cover cases like those:

```
Coin::from_str("ucosm") // no amount
Coin::from_str("") // empty input
Coin::from_str("-1ucosm") // negative amount
Coin::from_str(" 1ucosm") // unsupported whitespace
Coin::from_str("�1ucosm") // other broken data
```

Reply:
So, basically, there would be another error variant, something like "MissingAmount" that would throw in all those cases?

Reply:
I was thinking about using InvalidAmount for the new check, but given the error payload type and the separate check time, it is probably better to have `MissingAmount` or `AmountNotFound` separately.

Reply:
I added a check for this now. Only the empty one still throws `MissingDenom`, because that way the check becomes much more elegant (and I think both errors make sense here)